Source/WebKit/Platform/cocoa/LayerHostingContext.mm

```diff
@@ -170,6 +183,48 @@
 }
 
 #if USE(EXTENSIONKIT)
+#if ENABLE(MACH_PORT_LAYER_HOSTING)
+SafeSendRight LayerHostingContext::safeSendRight() const
+{
+    __block MachSendRight sendRight;
+    __block RetainPtr<NSData> dataRepresentation;
+    [[m_hostable handle] encodeWithBlock:^(mach_port_t copiedPort, NSData * _Nonnull data) {
+        sendRight = MachSendRight::create(copiedPort);
+        dataRepresentation = data;
+    }];
+    return { WTFMove(sendRight), FixedVector(makeVector(dataRepresentation.get())) };
+}
+
+RetainPtr<BELayerHierarchyHostingTransactionCoordinator> LayerHostingContext::createHostingUpdateCoordinator(const WebCore::SafeSendRight& safeSendRight)
+{
+    NSError *error = nil;
+    auto coordinator = [BELayerHierarchyHostingTransactionCoordinator coordinatorWithPort:safeSendRight.sendRight.sendRight() data:toNSData(safeSendRight.data.span()).get() error:&error];
+    if (error)
+        RELEASE_LOG_ERROR(Process, "Could not create update coordinator, error = %@", error);
+    return coordinator;
+}
+
+WebCore::SafeSendRight LayerHostingContext::fence(BELayerHierarchyHostingTransactionCoordinator *coordinator)
+{
+    __block MachSendRight sendRight;
+    __block RetainPtr<NSData> dataRepresentation;
+    [coordinator encodeWithBlock:^(mach_port_t copiedPort, NSData * _Nonnull data) {
+        sendRight = MachSendRight::create(copiedPort);
+        dataRepresentation = data;
+    }];
+    return { WTFMove(sendRight), FixedVector(makeVector(dataRepresentation.get())) };
+}
+
+RetainPtr<BELayerHierarchyHandle> LayerHostingContext::createHostingHandle(SafeSendRight&& safeSendRight)
+{
+    // We are leaking the send right here, since [BELayerHierarchyHandle handleWithPort] takes ownership of the send right.
+    NSError *error = nil;
+    auto handle = [BELayerHierarchyHandle handleWithPort:safeSendRight.sendRight.leakSendRight() data:toNSData(safeSendRight.data.span()).get() error:&error];
+    if (error)
```

Comment:
Does the send right get released, even in the error case?

Reply:
That is a good point! I believe the API implementation will also release the send right also in the error case. I will double-check this.

Thanks for reviewing!